File: ios/App/App/Plugins/NativeKakaoAuthPlugin.swift
import Capacitor
import Foundation
import KakaoSDKAuth
import KakaoSDKCommon
import KakaoSDKUser
import os.log

/**
 * Kakao Native Login — 카카오톡 우선, talk 실패(취소 제외) 시 카카오 계정(SDK WebView).
 * Chrome / Safari / 외부 브라우저 OAuth 금지.
 */
@objc(NativeKakaoAuthPlugin)
public class NativeKakaoAuthPlugin: CAPPlugin, CAPBridgedPlugin {
  private static let log = OSLog(subsystem: "com.dibay.app", category: "DIBAY_Kakao")

  public let identifier = "NativeKakaoAuthPlugin"
  public let jsName = "NativeKakaoAuth"
  public let pluginMethods: [CAPPluginMethod] = [
    CAPPluginMethod(name: "signIn", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name: "signOut", returnType: CAPPluginReturnPromise),
  ]

  private var pendingCall: CAPPluginCall?

  private func logEvent(_ event: String) {
    os_log("%{public}@", log: Self.log, type: .info, event)
  }

  private func rejectPending(code: String, message: String) {
    guard let call = pendingCall else { return }
    pendingCall = nil
    call.reject(code, message)
  }

  private func isValidKakaoAppKey(_ raw: String) -> Bool {
    let key = raw.trimmingCharacters(in: .whitespacesAndNewlines)
    if key.isEmpty { return false }
    if key.hasPrefix("$(") { return false }
    if key.contains("YOUR_KAKAO") { return false }
    return true
  }

  private func resolvePending(_ result: JSObject) {
    guard let call = pendingCall else { return }
    pendingCall = nil
    call.resolve(result)
  }

  private func isUserCancelled(_ error: Error) -> Bool {
    if let clientError = error as? ClientError, clientError.reason == .Cancelled {
      return true
    }
    return error.localizedDescription.lowercased().contains("cancel")
  }

  private func logFailure(_ error: Error) {
    if let clientError = error as? ClientError {
      logEvent("kakao_native_failed \(error.localizedDescription) cause=\(String(describing: clientError.reason))")
      return
    }
    logEvent("kakao_native_failed \(error.localizedDescription)")
  }

  private func rejectLoginError(_ error: Error) {
    logFailure(error)
    let message = error.localizedDescription
    let lower = message.lowercased()
    if lower.contains("kakaotalk is installed") || lower.contains("keyhash") || lower.contains("key hash") {
      rejectPending(code: "kakao_native_key_hash_required", message: message)
      return
    }
    rejectPending(code: "kakao_native_config_error", message: message)
  }

  private func finishLogin(token: OAuthToken?) {
    guard pendingCall != nil else { return }

    guard let token = token, !token.accessToken.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
      logEvent("kakao_native_token_missing")
      rejectPending(code: "kakao_native_token_missing", message: "Kakao access token missing")
      return
    }

    var result = JSObject()
    result["provider"] = "kakao"
    result["accessToken"] = token.accessToken
    if let refresh = token.refreshToken, !refresh.isEmpty {
      result["refreshToken"] = refresh
    }
    if let idToken = token.idToken, !idToken.isEmpty {
      result["idToken"] = idToken
    }

    UserApi.shared.me { [weak self] userResult, meError in
      guard let self = self else { return }
      guard self.pendingCall != nil else { return }

      if meError == nil, let user = userResult?.id {
        result["userId"] = String(user)
      }
      self.logEvent("kakao_native_success")
      self.resolvePending(result)
    }
  }

  private func handleToken(_ token: OAuthToken?, error: Error?) {
    guard pendingCall != nil else { return }

    if let error = error {
      if isUserCancelled(error) {
        logEvent("kakao_native_cancelled")
        rejectPending(code: "user_cancelled", message: "User cancelled Kakao sign-in")
        return
      }
      rejectLoginError(error)
      return
    }

    finishLogin(token: token)
  }

  @objc func signIn(_ call: CAPPluginCall) {
    logEvent("kakao_native_started")

    if pendingCall != nil {
      call.reject("kakao_native_in_flight", "Another Kakao sign-in is already in progress")
      return
    }

    guard let appKey = Bundle.main.object(forInfoDictionaryKey: "KAKAO_NATIVE_APP_KEY") as? String,
      isValidKakaoAppKey(appKey)
    else {
      call.reject("kakao_native_config_error", "KAKAO_NATIVE_APP_KEY is not configured")
      return
    }

    let trimmedKey = appKey.trimmingCharacters(in: .whitespacesAndNewlines)
    if !KakaoSDK.isInitialized() {
      KakaoSDK.initSDK(appKey: trimmedKey)
    }

    pendingCall = call

    if UserApi.isKakaoTalkLoginAvailable() {
      logEvent("kakao_native_talk_login")
      UserApi.shared.loginWithKakaoTalk { token, error in
        if self.pendingCall == nil {
          return
        }
        if let error = error, !self.isUserCancelled(error) {
          self.logEvent("kakao_native_talk_fallback_account")
          self.logEvent("kakao_native_account_login")
          UserApi.shared.loginWithKakaoAccount { token, error in
            self.handleToken(token, error: error)
          }
          return
        }
        self.handleToken(token, error: error)
      }
    } else {
      logEvent("kakao_native_account_login")
      UserApi.shared.loginWithKakaoAccount { token, error in
        self.handleToken(token, error: error)
      }
    }
  }

  @objc func signOut(_ call: CAPPluginCall) {
    UserApi.shared.logout { error in
      if error != nil {
        self.logEvent("kakao_native_signout_failed")
      } else {
        self.logEvent("kakao_native_signout_ok")
      }
      call.resolve()
    }
  }
}

File: android/app/src/main/java/com/dibay/app/NativeKakaoAuthPlugin.java
package com.dibay.app;

import android.app.Activity;
import android.util.Log;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.kakao.sdk.auth.model.OAuthToken;
import com.kakao.sdk.common.model.ClientError;
import com.kakao.sdk.common.model.ClientErrorCause;
import com.kakao.sdk.user.UserApiClient;
import kotlin.Unit;
import kotlin.jvm.functions.Function2;

/**
 * Kakao Native Login — 카카오톡 앱 우선, talk 실패(취소 제외) 시 카카오 계정(SDK WebView).
 * Chrome / Custom Tab / 외부 브라우저 OAuth 금지.
 */
@CapacitorPlugin(name = "NativeKakaoAuth")
public class NativeKakaoAuthPlugin extends Plugin {

  private static final String TAG = "DIBAY_Kakao";

  private PluginCall pendingCall;

  private void logEvent(String event) {
    Log.i(TAG, event);
  }

  private void rejectPendingCall(String code, String message) {
    PluginCall call = pendingCall;
    pendingCall = null;
    if (call != null) {
      call.reject(code, message);
    }
  }

  private boolean isUserCancelled(Throwable error) {
    if (error instanceof ClientError) {
      ClientErrorCause cause = ((ClientError) error).getReason();
      return cause == ClientErrorCause.Cancelled;
    }
    String message = error != null ? String.valueOf(error.getMessage()) : "";
    return message.toLowerCase().contains("cancel");
  }

  private void logFailure(Throwable error) {
    if (error == null) {
      return;
    }
    String message = error.getMessage() != null ? String.valueOf(error.getMessage()) : error.getClass().getSimpleName();
    if (error instanceof ClientError) {
      logEvent("kakao_native_failed " + message + " cause=" + ((ClientError) error).getReason().name());
      return;
    }
    logEvent("kakao_native_failed " + message);
  }

  private void rejectLoginError(Throwable error) {
    logFailure(error);
    String message = error.getMessage() != null ? String.valueOf(error.getMessage()) : "";
    String lower = message.toLowerCase();
    if (lower.contains("kakaotalk is installed") || lower.contains("keyhash") || lower.contains("key hash")) {
      rejectPendingCall("kakao_native_key_hash_required", message);
      return;
    }
    rejectPendingCall("kakao_native_config_error", message);
  }

  private Function2<OAuthToken, Throwable, Unit> loginCallback = (token, error) -> {
    PluginCall call = pendingCall;
    if (call == null) {
      return Unit.INSTANCE;
    }

    if (error != null) {
      if (isUserCancelled(error)) {
        logEvent("kakao_native_cancelled");
        rejectPendingCall("user_cancelled", "User cancelled Kakao sign-in");
      } else {
        rejectLoginError(error);
      }
      return Unit.INSTANCE;
    }

    if (token == null || token.getAccessToken() == null || token.getAccessToken().trim().isEmpty()) {
      logEvent("kakao_native_token_missing");
      rejectPendingCall("kakao_native_token_missing", "Kakao access token missing");
      return Unit.INSTANCE;
    }

    JSObject result = new JSObject();
    result.put("provider", "kakao");
    result.put("accessToken", token.getAccessToken());
    if (token.getRefreshToken() != null && !token.getRefreshToken().trim().isEmpty()) {
      result.put("refreshToken", token.getRefreshToken());
    }
    if (token.getIdToken() != null && !token.getIdToken().trim().isEmpty()) {
      result.put("idToken", token.getIdToken());
    }

    UserApiClient.getInstance().me((user, meError) -> {
      if (pendingCall != call) {
        if (call != null) {
          call.reject("kakao_native_unavailable", "Kakao sign-in session changed");
        }
        return Unit.INSTANCE;
      }
      if (meError != null) {
        rejectPendingCall("kakao_native_config_error", meError.getMessage() != null ? meError.getMessage() : "Kakao profile fetch failed");
        return Unit.INSTANCE;
      }
      if (user != null && user.getId() != null) {
        result.put("userId", String.valueOf(user.getId()));
      }
      logEvent("kakao_native_success");
      pendingCall = null;
      call.resolve(result);
      return Unit.INSTANCE;
    });

    return Unit.INSTANCE;
  };

  @Override
  protected void handleOnDestroy() {
    // DO NOT reject or clear pendingCall — Kakao SDK may complete after Activity recreate.
    super.handleOnDestroy();
  }

  @PluginMethod
  public void signIn(PluginCall call) {
    logEvent("kakao_native_started");

    if (pendingCall != null) {
      call.reject("kakao_native_in_flight", "Another Kakao sign-in is already in progress");
      return;
    }

    String appKey = getContext().getString(R.string.kakao_native_app_key).trim();
    if (appKey.isEmpty()) {
      call.reject("kakao_native_config_error", "KAKAO_NATIVE_APP_KEY is not configured");
      return;
    }

    Activity activity = getActivity();
    if (activity == null) {
      call.reject("kakao_native_unavailable", "Activity not found");
      return;
    }

    pendingCall = call;

    if (UserApiClient.getInstance().isKakaoTalkLoginAvailable(activity)) {
      startKakaoTalkLogin(activity);
    } else {
      startKakaoAccountLogin(activity);
    }
  }

  /** 카카오 공식 샘플: talk 실패(취소 제외) 시 account 로그인으로 자동 전환 */
  private void startKakaoTalkLogin(Activity activity) {
    logEvent("kakao_native_talk_login");
    UserApiClient.getInstance().loginWithKakaoTalk(activity, (token, error) -> {
      if (pendingCall == null) {
        return Unit.INSTANCE;
      }
      if (error != null && !isUserCancelled(error)) {
        logEvent("kakao_native_talk_fallback_account");
        startKakaoAccountLogin(activity);
        return Unit.INSTANCE;
      }
      return loginCallback.invoke(token, error);
    });
  }

  private void startKakaoAccountLogin(Activity activity) {
    if (pendingCall == null) {
      return;
    }
    Activity active = activity != null ? activity : getActivity();
    if (active == null) {
      rejectPendingCall("kakao_native_unavailable", "Activity not found");
      return;
    }
    logEvent("kakao_native_account_login");
    UserApiClient.getInstance().loginWithKakaoAccount(active, loginCallback);
  }

  @PluginMethod
  public void signOut(PluginCall call) {
    UserApiClient.getInstance().logout(error -> {
      if (error != null) {
        logEvent("kakao_native_signout_failed");
      } else {
        logEvent("kakao_native_signout_ok");
      }
      call.resolve();
      return Unit.INSTANCE;
    });
  }
}
